### PR TITLE
Implement initial BindLeft and MonadLeft

### LINKED ISSUE #12: Add a `BindLeft` type class
```purescript
class BindLeft m where
  lbind :: forall a b r. m a r -> (a -> m b r) -> m b r

class (Biapplicative m, BindLeft m) <= MonadLeft m
```

This is a generalized, type-changing `catch` operation, which is useful for things like `Either`, and also lets you implement a `Monad` instance for `Flip`.

### PR DESCRIPTION
**Pull request by JordanMartinez**
Fixes #12

WIP

Things to do:
- [x] Implement qualified do for `ldiscard`, `lbind`, `lmap`, and `lapply`

### CODE REVIEW

File: src/Control/MonadLeft/Qualified.purs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+-- | Enables one to use `ado notation` and `do notation` on the left parameter
+-- | of a Bifunctor.
+-- |
+-- | `ado notation` example
+-- | ```
+-- | import Control.MonadLeft.Qualified as MonadLeft
+-- |
+-- | foo :: Either Int String -> Either String String
+-- | foo comp = MonadLeft.ado
```
Comment (JordanMartinez) [inline src/Control/MonadLeft/Qualified.purs:9]: Should the convention be `Flip.do` rather than `MonadLeft.do`?

**Comment by JordanMartinez** [comment]
What advantage would there be for defining an instance of `MonadLeft` and its superclasses instances for `Flip` if we have qualified do?

File: src/Control/MonadLeft/Qualified.purs (hunk 1 above)
Comment (kl0tl) [reply src/Control/MonadLeft/Qualified.purs:9]: `MonadLeft.do` seems more meaningful to me than `Flip.do`, but writing `MonadLeft.ado` for a type with only an `ApplicativeLeft` instance and no `MonadLeft` is odd.

File: src/Control/MonadLeft/Qualified.purs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+-- | ```
+-- | import Control.MonadLeft.Qualified as MonadLeft
+-- |
+-- | foo :: Either Int String -> Either String String
+-- | foo comp = MonadLeft.do
+-- |   a <- comp
+-- |   b <- comp
+-- |   lpure $ show $ a + b
+-- | ```
+module Control.MonadLeft.Qualified where
```
Comment (kl0tl) [inline src/Control/MonadLeft/Qualified.purs:24]: Don’t we also need to export `pure` for Applicative do notation?

**Review by kl0tl — COMMENTED** [review]
While `lapply`, `lpure`, `lbind` and `ldiscard` are consistent with `lmap`, `llift*` is awkward and `lifM` reads like a typo. This is consistent with the prefixes from bifunctors and indexed-monad though, so perhaps this ship has already sailed.

**Comment by JordanMartinez** [comment]
> This is consistent with the prefixes from bifunctors and indexed-monad though, so perhaps this ship has already sailed.

We could always break convention here now and update bifunctors and indexed-monad to use a `*Left` suffix in a future breaking changes release. In this case, I'd rather have something readable than something short.

File: src/Control/MonadLeft/Qualified.purs (hunk 2 above)
Comment (JordanMartinez) [reply src/Control/MonadLeft/Qualified.purs:24]: I considered `Left.do`, but wasn't sure whether that would clash with the `Left` constructor for `Either`. `Flip` has some similarity to the `Flip` newtype, but it won't prevent clashing either. `MonadLeft` seems a bit much to type when other usages (e.g. `Ix`, `React`, `Hooks`, etc.) are shorter.

What about `BiLeft.do`? `Bi` for bifunctor and `Left` to indicate which parameter it's working on? I'm not sure whether every MonadLeft will have also be a Bifunctor... Maybe `MLeft` as a shortened version of `MonadLeft`?

Comment (JordanMartinez) [reply src/Control/MonadLeft/Qualified.purs:24]: > Don’t we also need to export pure for Applicative do notation?

I don't believe so. Any usage of `pure` will likely be prefixed with the qualified do name. For example:
```purescript
main = Ix.do
  foo <- Ix.pure "foo"
  -- ...
```

File: src/Control/ApplyLeft.purs [hunk 3]
```diff
@@ -0,0 +1,36 @@
+module Control.ApplyLeft where
+
+import Prelude
+import Data.Bifunctor (class Bifunctor, lmap)
+
+-- | Same as `Apply` but works on the left parameter in a Bifunctor.
+class Bifunctor w <= ApplyLeft w where
```
[the remark this answers is not in the document]
Comment (JordanMartinez) [reply src/Control/ApplyLeft.purs:7]: Pretty sure I copied and pasted this from somewhere and then updated it. So, it should be updated to `m`.

File: src/Control/BindLeft.purs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import Data.Unit (Unit)
+import Control.Category (identity)
+
+-- | Same as `Discard` but works on the left parameter of the bifunctor
+-- | rather than the right parameter, which is the default for `Discard`.
+class DiscardLeft a where
+  ldiscard :: forall k (m :: Type -> k -> Type) b r. BindLeft m => m a r -> (a -> m b r) -> m b r
+
+instance discardLeftUnit :: DiscardLeft Unit where
+  ldiscard ma aToMB = lbind ma aToMB
```
[the remark this answers is not in the document]
Comment (JordanMartinez) [reply src/Control/BindLeft.purs]: Fixed in latest commits

File: src/Control/BindLeft.purs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  lbind :: forall a b r. m a r -> (a -> m b r) -> m b r
+
+ljoin :: forall m a r. BindLeft m => m (m a r) r -> m a r
+ljoin m = lbind m identity
+
+lcomposeKleisli :: forall m a b c r. BindLeft m => (a -> m b r) -> (b -> m c r) -> a -> m c r
+lcomposeKleisli aToMB bToMC a = lbind (aToMB a) bToMC
+
+lcomposeKleisliFlipped :: forall m a b c r. BindLeft m => (b -> m c r) -> (a -> m b r) -> a -> m c r
+lcomposeKleisliFlipped bToMC aToMB a = lbind (aToMB a) bToMC
```
[the remark this answers is not in the document]
Comment (JordanMartinez) [reply src/Control/BindLeft.purs]: Fixed in latest commits

File: src/Control/ApplicativeLeft.purs [hunk 6]
```diff
@@ -0,0 +1,18 @@
+module Control.ApplicativeLeft where
+
+import Prelude
+import Control.ApplyLeft (class ApplyLeft, lapply)
+
+-- | Same as `Applicative` but works on the left parameter in a Bifunctor.
+class ApplyLeft w <= ApplicativeLeft w where
```
[the remark this answers is not in the document]
Comment (JordanMartinez) [reply src/Control/ApplicativeLeft.purs]: Fixed in latest commits

File: src/Control/MonadLeft/Qualified.purs (hunk 1 above)
Comment (JordanMartinez) [reply src/Control/MonadLeft/Qualified.purs:9]: I read `MonadLeft.do` / `MonadLeft.ado` less of a "hey, this thing is a monad!" and more of a "hey, we're using the `MonadLeft` type class hierarchy here. That's how I read `Ix.do`, `React.do`, and `Hooks.do`.

**Comment by JordanMartinez** [comment]
CI now builds.

**Comment by JordanMartinez** [comment]
The only thing blocking this PR is deciding on the naming convention for referencing this qualified do. I think `BiLeft` should be used as it's short for `bifunctor` and the `left` parameter used in the bifunctor.

**Comment by JordanMartinez** [comment]
I moved all the modules implemented here into a `Control.Bifunctor` module. This is so we reduce possible breaking changes if we ever want to implement a `MonadLeft`-like type class hierarchy for trifunctors. 

I've also updated the naming convention for the qualified do to use `BiLeft` which distinguishes it from the `Left` constructor of Either and ties it to the underlying bifunctor type's left type parameter.

Can I get an approval on this?

**Comment by hdgarrood** [comment]
> This is so we reduce possible breaking changes if we ever want to implement a MonadLeft-like type class hierarchy for trifunctors.

This feels like a case of YAGNI to me. It's hard for me to see us ever wanting to do that in core; even this version of MonadLeft has only generated minimal interest.

**Comment by hdgarrood** [comment]
I don't have a strong view on whether MonadLeft should be in core or not, but I think the argument that it should be here so that we can provide `Monad Flip` makes sense to me. It's just that I think it probably is safe to say that a version of MonadLeft which works on the first of _three_ type variables is not useful often enough to deserve to be in core, and therefore I'd suggest not designing this library in such a way as to make space to do that in the future.

**Comment by JordanMartinez** [comment]
Ah... That does make sense.

**Comment by JordanMartinez** [comment]
Closing as this hasn't received an approval or further feedback.

**Pull request closed without merging**